lib/clients/ZookeeperManager.js

```diff
@@ -0,0 +1,313 @@
+                // getData() instead
+                this.getData('/', err => {
+                    if (err && err.name !== 'NO_NODE') {
+                        this.emit('error', err);
+                        return;
+                    }
+                    // NO_NODE error and autoCreateNamespace is enabled
+                    if (err) {
+                        const nsIndex = this.connectionString.indexOf('/');
+                        const hostPort = this.connectionString.slice(0, nsIndex);
+                        const namespace = this.connectionString.slice(nsIndex);
+                        const rootZkClient = zookeeper.createClient(hostPort, this.options);
+                        rootZkClient.connect();
+                        rootZkClient.mkdirp(namespace, err => {
+                            if (err && err.name !== 'NODE_EXISTS') {
+                                this.emit('error');
+                                return;
+                            }
+                            this.emit('ready');
+                            return;
+                        });
+                        return;
+                    }
+                    this.emit('ready');
+                    return;
+                });
+                return;
+            }
+            this.emit('ready');
+            return;
+        });
+
+        this.client.once('expired', () => {
+            this.log.info('zookeeper client expired', {
+                method: 'ZookeeperManager.once.expired',
+            });
+            // close and clean up the existing ZooKeeper client connection.
+            this.close();
+            // establish a new session with the ZooKeeper.
+            this._connect();
+        });
+
+        this.client.on('state', (state) => {
+            this.log.debug('zookeeper new state', {
+                state,
+                method: 'ZookeeperManager.on.state',
+            });
+        });
+
+        // Forward ZooKeeper events
+        this.client.on('connected', () => {
+            this.emit('connected');
+        });
+
+        this.client.on('disconnected', () => {
+            this.emit('disconnected');
+        });
+
+        this.client.connect();
+    }
+
+    // Forward ZooKeeper methods
+
+    /**
+     * Shutdown the client.
+     * @method close
+     * @return {undefined}
+     */
+    close() {
+        this.client.close();
+        return;
+    }
+
+    /**
+     * Create a node with given path, data, acls and mode.
+     *
+     * @method create
+     * @param {String} path The node path.
+     * @param {Buffer} [data=undefined] The data buffer.
+     * @param {Array} [acls=ACL.OPEN_ACL_UNSAFE] An array of ACL object.
+     * @param {CreateMode} [mode=CreateMode.PERSISTENT] The creation mode.
+     * @param {Function} callback The callback function.
+     * @return {undefined}
+     */
+    create(path, data, acls, mode, callback) {
```

Comment:
I'm not too fond of having to reproduce the original API of the zookeeper client, for the maintenance burden notably if we have to use new functions later and for the extra code.

I am thinking, would it be better to extend the zookeeper client itself in such case, as if we're manipulating the zookeeper client but with extra logic inside to deal with session expiration etc., possibly it would cause some issues on its own, so not sure. Not blocking for me but just mentioning it doesn't look like an ideal solution.